crates/iceberg/src/metadata_scan.rs

```diff
@@ -0,0 +1,407 @@
+pub struct MetadataLogEntriesTable;
+
+impl MetadataLogEntriesTable {
+    fn snapshot_id_as_of_time(
+        table_metadata: &TableMetadataRef,
+        timestamp_ms_inclusive: i64,
+    ) -> Option<&SnapshotRef> {
+        let mut snapshot_id = None;
+        // The table metadata snapshot log is chronological
+        for log_entry in table_metadata.history() {
+            if log_entry.timestamp_ms <= timestamp_ms_inclusive {
+                snapshot_id = Some(log_entry.snapshot_id);
+            }
+        }
+        snapshot_id.and_then(|id| table_metadata.snapshot_by_id(id))
+    }
+}
+
+impl MetadataTable for MetadataLogEntriesTable {
+    fn schema() -> Schema {
+        Schema::new(vec![
+            Field::new(
+                "timestamp",
+                DataType::Timestamp(TimeUnit::Millisecond, Some("+00:00".into())),
+                false,
+            ),
+            Field::new("file", DataType::Utf8, false),
+            Field::new("latest_snapshot_id", DataType::Int64, true),
+            Field::new("latest_schema_id", DataType::Int32, true),
+            Field::new("latest_sequence_number", DataType::Int64, true),
+        ])
+    }
+
+    fn scan(scan: &MetadataScan) -> Result<RecordBatch> {
+        let mut timestamp =
+            PrimitiveBuilder::<TimestampMillisecondType>::new().with_timezone("+00:00");
+        let mut file = StringBuilder::new();
+        let mut latest_snapshot_id = PrimitiveBuilder::<Int64Type>::new();
+        let mut latest_schema_id = PrimitiveBuilder::<Int32Type>::new();
+        let mut latest_sequence_number = PrimitiveBuilder::<Int64Type>::new();
+
+        let mut append_metadata_log_entry = |timestamp_ms: i64, metadata_file: &str| {
+            timestamp.append_value(timestamp_ms);
+            file.append_value(metadata_file);
+
+            let snapshot =
+                MetadataLogEntriesTable::snapshot_id_as_of_time(&scan.metadata_ref, timestamp_ms);
+            latest_snapshot_id.append_option(snapshot.map(|s| s.snapshot_id()));
+            latest_schema_id.append_option(snapshot.and_then(|s| s.schema_id()));
+            latest_sequence_number.append_option(snapshot.map(|s| s.sequence_number()));
+        };
+
+        for metadata_log_entry in scan.metadata_ref.metadata_log() {
+            append_metadata_log_entry(
+                metadata_log_entry.timestamp_ms,
+                &metadata_log_entry.metadata_file,
+            );
+        }
+
+        // Include the current metadata locaction and modification time in the table. This matches
+        // the Java implementation. Unlike the Java implementation, a current metadata location is
+        // optional here. In that case, we omit current metadata from the metadata log table.
+        if let Some(current_metadata_location) = &scan.metadata_location {
+            append_metadata_log_entry(
+                scan.metadata_ref.last_updated_ms(),
+                current_metadata_location,
+            );
+        }
```

Comment:
Linking [Java] and [Python] implementations. As commented, the difference is that the metadata location is optional in `Table`.

Alternatively, we could `expect` here.

[Java]: https://github.com/apache/iceberg/blob/8a70fe0ff5f241aec8856f8091c77fdce35ad256/core/src/main/java/org/apache/iceberg/MetadataLogEntriesTable.java#L62-L66
[Python]: https://github.com/apache/iceberg-python/blob/0e5086ceb77351bc0b6ec3a592f5eda70a0afe46/pyiceberg/table/inspect.py#L438-L442